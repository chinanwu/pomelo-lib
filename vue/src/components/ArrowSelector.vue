<template>
  <div class="ArrowSelector">
    <div class="ArrowSelector__main">
      <button
        id="arrowSelectorLeft"
        class="ArrowSelector__btn"
        @click="handleLeftArrow"
        @keydown="handleLeftArrowKeyDown"
      >
        &lt;
      </button>
      <span>{{ current }} / {{ total }}</span>
      <button
        id="arrowSelectorRight"
        class="ArrowSelector__btn"
        @click="handleRightArrow"
        @keydown="handleRightArrowKeyDown"
      >
        &gt;
      </button>
    </div>
    <div class="ArrowSelector__label">Current</div>
  </div>
</template>

<script>
export default {
  name: "ArrowSelector",
  props: {
    default: {
      type: Number,
      required: true,
      validator: value => value > 0
    },
    total: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      current: this.default
    };
  },
  methods: {
    setCurrent(next) {
      this.current = next;
      this.$emit("change", next);
    },
    handleLeftArrow() {
      this.setCurrent(this.current - 1 <= 0 ? this.total : this.current - 1);
    },
    handleRightArrow() {
      this.setCurrent(this.current + 1 > this.total ? 1 : this.current + 1);
    },
    handleLeftArrowKeyDown(event) {
      if (
        event &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        event.key
      ) {
        if (event.key === "Enter" || event.key === "Space") {
          event.preventDefault();
          this.handleLeftArrow();
        }
      }
    },
    handleRightArrowKeyDown(event) {
      if (
        event &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        event.key
      ) {
        if (event.key === "Enter" || event.key === "Space") {
          event.preventDefault();
          this.handleRightArrow();
        }
      }
    }
  },
  emits: ["change"]
};
</script>

<style lang="less" scoped>
.ArrowSelector {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ArrowSelector__main {
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
  font-size: 3.6rem;
}

.ArrowSelector__btn {
  background-color: inherit;
  border: none;
  font-size: 7.2rem;
  cursor: pointer;
  color: #c8c8c8;
}

.ArrowSelector__label {
  font-size: 2.4rem;
  color: #4f4f4f;

  &::after {
    content: " Project";
  }
}

// iPhone 11 Pro Max, iPhone 8
@media only screen and (max-width: 414px) {
  .ArrowSelector__main {
    font-size: 3.6rem;
  }

  .ArrowSelector__btn {
    font-size: 4.8rem;
  }

  .ArrowSelector__label::after {
    content: "";
  }
}

// iPhone SE
@media only screen and (max-width: 320px) {
  .ArrowSelector__main {
    font-size: 2.4rem;
  }

  .ArrowSelector__btn {
    font-size: 4.8rem;
  }

  .ArrowSelector__label {
    font-size: 1.8rem;
  }
}
</style>
