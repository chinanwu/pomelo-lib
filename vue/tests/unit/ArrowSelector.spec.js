import { shallowMount } from "@vue/test-utils";

import ArrowSelector from "@/components/ArrowSelector";

describe("ArrowSelector", () => {
  describe("rendering", () => {
    it("renders ArrowSelector", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 1,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);
    });
  });

  describe("behaviour", () => {
    it("selects next if click right arrow", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 1,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);

      wrapper.find("#arrowSelectorRight").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[2]]);
      expect(wrapper.vm.current).toBe(2);
    });

    it("selects next if enter press right arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 3,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Enter",
        preventDefault
      });
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[1]]);
      expect(wrapper.vm.current).toBe(1);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("selects next if space press right arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 3,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Space",
        preventDefault
      });
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[1]]);
      expect(wrapper.vm.current).toBe(1);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("does nothing on invalid keydown on right arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 3,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "A",
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorRight").trigger("keydown");
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Enter",
        shiftKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Enter",
        ctrlKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Enter",
        altKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorRight").trigger("keydown", {
        key: "Enter",
        metaKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);
    });

    it("selects prev if click left arrow", () => {
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 2,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(2);

      wrapper.find("#arrowSelectorLeft").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[1]]);
      expect(wrapper.vm.current).toBe(1);
    });

    it("selects prev if enter press left arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 1,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Enter",
        preventDefault
      });
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[3]]);
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("selects prev if space press left arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 1,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(1);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Space",
        preventDefault
      });
      expect(wrapper.emitted()).toHaveProperty("change");
      expect(wrapper.emitted().change).toEqual([[3]]);
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("does nothing on invalid keydown on left arrow", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(ArrowSelector, {
        props: {
          default: 3,
          total: 3
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "A",
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorLeft").trigger("keydown");
      expect(wrapper.vm.current).toBe(3);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Enter",
        shiftKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Enter",
        ctrlKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Enter",
        altKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);

      wrapper.find("#arrowSelectorLeft").trigger("keydown", {
        key: "Enter",
        metaKey: true,
        preventDefault
      });
      expect(wrapper.vm.current).toBe(3);
      expect(preventDefault).toHaveBeenCalledTimes(0);
    });
  });

  describe("prop validation", () => {
    it("throws error if not valid default", () => {
      const validator = ArrowSelector.props.default.validator;
      expect(validator(0)).toBe(false);
      expect(validator(1)).toBe(true);
    });
  });
});
