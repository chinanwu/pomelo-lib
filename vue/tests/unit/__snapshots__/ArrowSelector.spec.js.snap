// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowSelector behaviour does nothing on invalid keydown on left arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      3 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour does nothing on invalid keydown on right arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      3 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects next if click right arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      1 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects next if enter press right arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      3 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects next if space press right arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      3 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects prev if click left arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      2 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects prev if enter press left arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      1 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector behaviour selects prev if space press left arrow 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      1 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;

exports[`ArrowSelector rendering renders ArrowSelector 1`] = `
<div
  class="ArrowSelector"
>
  <div
    class="ArrowSelector__main"
  >
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorLeft"
    >
       &lt; 
    </button>
    <span>
      1 / 3
    </span>
    <button
      class="ArrowSelector__btn"
      id="arrowSelectorRight"
    >
       &gt; 
    </button>
  </div>
  <div
    class="ArrowSelector__label"
  >
    Current
  </div>
</div>
`;
